crates/subspace-service/src/sync_from_dsn/snap_sync.rs

```diff
@@ -95,55 +97,85 @@ pub(crate) async fn snap_sync<Backend, Block, AS, Client, PG, NR>(
         .await;
 }
 
-#[allow(clippy::too_many_arguments)]
-async fn sync<PG, AS, Block, Client, IQS, B, NR>(
+// Get blocks from the last segment or from the segment containing the target block.
+// Returns encoded blocks collection and used segment index.
+pub(crate) async fn get_blocks_from_target_segment<AS, PG>(
     segment_headers_store: &SegmentHeadersStore<AS>,
     node: &Node,
     piece_getter: &PG,
-    fork_id: Option<&str>,
-    client: &Arc<Client>,
-    import_queue_service: &mut IQS,
-    network_request: &NR,
-    sync_service: &SyncingService<Block>,
-) -> Result<(), Error>
+    target_block: Option<BlockNumber>,
+) -> Result<Option<(SegmentIndex, VecDeque<(BlockNumber, Vec<u8>)>)>, Error>
 where
-    B: sc_client_api::Backend<Block>,
-    PG: DsnSyncPieceGetter,
     AS: AuxStore,
-    Block: BlockT,
-    Client: HeaderBackend<Block>
-        + ClientExt<Block, B>
-        + ProvideRuntimeApi<Block>
-        + ProofProvider<Block>
-        + LockImportRun<Block, B>
-        + Send
-        + Sync
-        + 'static,
-    Client::Api: SubspaceApi<Block, FarmerPublicKey> + ObjectsApi<Block>,
-    IQS: ImportQueueService<Block> + ?Sized,
-    NR: NetworkRequest,
+    PG: DsnSyncPieceGetter,
 {
-    debug!("Starting snap sync...");
-
     sync_segment_headers(segment_headers_store, node)
         .await
         .map_err(|error| format!("Failed to sync segment headers: {}", error))?;
 
-    let last_segment_index = segment_headers_store
-        .max_segment_index()
-        .expect("Successfully synced above; qed");
+    let target_segment_index = {
+        let last_segment_index = segment_headers_store
+            .max_segment_index()
+            .expect("Successfully synced above; qed");
+
+        if let Some(target_block) = target_block {
+            let mut segment_header = segment_headers_store
+                .get_segment_header(last_segment_index)
+                .ok_or(format!(
+                    "Can't get segment header from the store: {last_segment_index}"
+                ))?;
+
+            if segment_header.last_archived_block().number < target_block {
+                return Err(format!(
+                    "Target block is greater than the last archived block. \
+                    Last segment index = {last_segment_index}, target block = {target_block}, \
+                    last block from the segment = {}
+                    ",
+                    segment_header.last_archived_block().number
```

Comment:
The error description is exactly the way I understand the code, but `if` condition is harder to read because it is inverted:
```suggestion
            if target_block > segment_header.last_archived_block().number {
                return Err(format!(
                    "Target block is greater than the last archived block. \
                    Last segment index = {last_segment_index}, target block = {target_block}, \
                    last block from the segment = {}",
                    segment_header.last_archived_block().number
```

While the logic is exactly the same either way, suggested version is easier to comprehend, we basically are looking for something beyond what is archived in the latest segment, so it is clearly out of reach. What you had before could be described like "block that has not yet been archived", but even then it is a bit harder to read IMO.

And I would then update checks below in a similar way.